### ISSUE TITLE
[BUG] Sharding is sub-optimal

### ISSUE THREAD

**Issue by myieye** [microsoft/playwright#19681]
**Context:**
- Playwright Version: 1.29.1
- Operating System: Windows
- Node.js version: 16.17.0
- Browser: Chromium

**Describe the bug**

I have 16 test files and as soon as I use more than 6 shards (e.g. 7 or 8), some shards are empty.
With more test files than shards, I would expect each shard to get at least some work to do.
I can't find any documentation that explains why this might make sense, so it just seems like a bug to me.

My tests:
![image](https://user-images.githubusercontent.com/12587509/209350353-36776fb3-d418-413c-a448-c70c5cf5488a.png)

Output:
![image](https://user-images.githubusercontent.com/12587509/209350414-08b1ce16-ea42-4353-bc7a-3d18014b9a5e.png)

Here's a GitHub action that failed, because it was suprised to not find any test output:
https://github.com/myieye/web-languageforge/actions/runs/3764843792/jobs/6399701097

Can be reproduced on this branch:
https://github.com/myieye/web-languageforge/tree/chore/run-e2e-tests-in-shards

**Comment by pavelfeldman** [comment]
There is hardly any point in using sharding unless you have hundreds of files - setting up a shard on CI would take more time than the benefit from the parallelization. We are considering various factors including shared fixtures to minimize the setup time, so sharding does not mean each shard gets exactly the same number of tests.

**Comment by myieye** [comment]
Feel free to close it if this is the expected behaviour.

Setting up shards in a GitHub workflow is trivial and is already saving us about 40min per build. So, I don't quite buy your argument.

**Comment by dgozman** [comment]
@myieye I think this is because Playwright does not run tests in a single file in parallel by default, and therefore cannot shard them since that we unexpected. You can explicitly opt-in into [`fullyParallel` mode](https://playwright.dev/docs/api/class-testconfig#test-config-fully-parallel) in your config, and that will most likely produce non-empty shards. Let me know if that helps.

**Comment by dgozman** [comment]
Closing as per above. If you still encounter the problem after the fix above, please file a new issue with a repro and link to this one.

**Comment by Skn0tt** [comment]
Hi folks! We're playing around with a new sharding feature that's useful for balancing test durations between different shards: https://github.com/microsoft/playwright/pull/38624
We'd appreciate you to try it out and leave us your thoughts in the PR.
